packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts

```diff
@@ -95,6 +96,17 @@ export class DockerImageAsset extends Construct implements assets.IAsset {
       exclude = [...exclude, ...fs.readFileSync(ignore).toString().split('\n').filter(e => !!e)];
     }
 
+    // .dockerignore files allow you to ignore the Dockerfile and .dockerignore file
+    // so that these files do not get sent to the Daemon during COPY or ADD. The
+    // files are still utilized by the Docker daemon, even if they are ignored
+    // We need to copy these files into the asset folder, even if they
+    // are included in the .dockerignore, so that these files can be used during the
+    // docker build, so remove any excludes that would match these two essential files
+    exclude = exclude.filter(ignoreExpression => {
+      return !(minimatch('Dockerfile', ignoreExpression, { dot: true }) ||
```

Comment:
What if users specify a custom name for the docker file. Also, why do we need minimatch here? Let’s just always copy the dockerfile to the staging area instead of patching up the exclude list

Reply:
Yes good catch, I will follow up with a feature to always add the custom name for the docker file.

For your second suggestion it looks like the copy to the staging area is happening here: https://github.com/aws/aws-cdk/blob/a50fd94aacc1c97e3e9e1e6405967d4795003bf1/packages/%40aws-cdk/aws-ecr-assets/lib/image-asset.ts#L127-L134

But I'm not seeing any method on the [`assets.Staging()` constructor](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_assets.Staging.html), or any method on the `Staging` class for copying a specific file into the staging area? It looks like my only options are to supply a single source folder path, and then a list of file globs to exclude, hence my approach was to filter out excluding file globs that would cause the required files to be excluded. I figured that would be the most minimal way to ensure that a required file doesn't get excluded.

What would be your recommended approach for ensuring that it a required file always makes it into the staging area? Should I extend the `assets.Staging()` with a new property which is a list of `includes`?

Maybe something like:

1. Copy the source path but excluding the files matching `excludes`
2. Copy the source path again but this time copy only files matching `includes`

Reply:
Alternative I could try to add a new method that works like this:

```
const staging = new assets.Staging(this, 'Staging', { 
   ...props, 
   exclude, 
   sourcePath: dir, 
   extraHash: Object.keys(extraHash).length === 0 
     ? undefined 
     : JSON.stringify(extraHash) 
 }); 

staging.addFiles(sourcePath, fileGlob);
```

However this appears like it would require a fairly significant change to the Staging class, to make it keep track of a list of sourcePaths and patterns, and then in the [synthesize method](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/assets/lib/staging.ts#L69-L91) loop through the list copying each into the asset folder.

Reply:
At first glance I was also leaning towards what @eladb suggested. 

But @nathanpeck is right, there is no straightforward way of explicitly adding files to the staging area, since the copy only happens during `synthesize`. And it seems that every attempt to enable this only makes the flow less clear. Im good with filtering the `excludes` list.

Regarding the custom docker file, yeah, this needs a fix. You can just use the already defined `file` const. 

Also - the fact the CI passed with this bug means we are missing a test. Could you add a test that uses a custom docker file in the ignore file?